x/oracle/keeper/native_token.go

```diff
@@ -35,6 +44,90 @@ func (k Keeper) GetStakerInfo(ctx sdk.Context, assetID, stakerAddr string) types
 	return stakerInfo
 }
 
+// TODO: pagination
+// GetStakerInfos returns all stakers information
+func (k Keeper) GetStakerInfos(ctx sdk.Context, assetID string) (ret []*types.StakerInfo) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, types.NativeTokenStakerKeyPrefix(assetID))
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		sInfo := types.StakerInfo{}
+		k.cdc.MustUnmarshal(iterator.Value(), &sInfo)
+		// keep only the latest effective-balance
+		sInfo.BalanceList = sInfo.BalanceList[:len(sInfo.BalanceList)-1]
+		// this is mainly used by price feeder, so we remove the stakerAddr to reduce the size of return value
+		sInfo.StakerAddr = ""
+		ret = append(ret, &sInfo)
+	}
+	return ret
+}
+
+// GetAllStakerInfosAssets returns all stakerInfos combined with assetIDs they belong to, used for genesisstate exporting
+func (k Keeper) GetAllStakerInfosAssets(ctx sdk.Context) (ret []types.StakerInfosAssets) {
+	store := ctx.KVStore(k.storeKey)
+	// set assetID as "" to iterate all value with different assetIDs
+	iterator := sdk.KVStorePrefixIterator(store, types.NativeTokenStakerKeyPrefix(""))
+	defer iterator.Close()
+	ret = make([]types.StakerInfosAssets, 0)
+	l := 0
+	for ; iterator.Valid(); iterator.Next() {
+		assetID, _ := types.ParseNativeTokenStakerKey(iterator.Key())
+		if l == 0 || ret[l-1].AssetId != assetID {
+			ret = append(ret, types.StakerInfosAssets{
+				AssetId:     assetID,
+				StakerInfos: make([]*types.StakerInfo, 0),
+			})
+			l++
+		}
+		v := &types.StakerInfo{}
+		k.cdc.MustUnmarshal(iterator.Value(), v)
+		ret[l-1].StakerInfos = append(ret[l-1].StakerInfos, v)
+	}
+	return ret
+}
+
+// SetStakerList set staker list for assetID, this is mainly used for genesis init
+func (k Keeper) SetStakerList(ctx sdk.Context, assetID string, sl *types.StakerList) {
+	if sl == nil {
+		return
+	}
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshal(sl)
+	store.Set(types.NativeTokenStakerListKey(assetID), bz)
+	// set cache used by updateNSTByBalanceChange
+	stakerList = *sl
```

Comment:
Not sure about this, but if the same function is called using EstimateGas or Call or CheckTx, won't this cause a mistake in the value?